reconcile/openshift_namespaces.py

```diff
@@ -1,127 +1,137 @@
 import logging
 
-import reconcile.utils.gql as gql
+from typing import List, Dict, Optional, Tuple, Any, Iterable, Mapping
 import reconcile.utils.threaded as threaded
-import reconcile.openshift_base as ob
 import reconcile.queries as queries
 
-from reconcile.utils.openshift_resource import ResourceInventory
 from reconcile.utils.oc import OC_Map
 from reconcile.utils.oc import StatusCodeError
 from reconcile.utils.defer import defer
 from reconcile.utils.sharding import is_in_shard
 
 
-QUERY = """
-{
-  namespaces: namespaces_v1 {
-    name
-    cluster {
-      name
-      serverUrl
-      jumpHost {
-          hostname
-          knownHosts
-          user
-          port
-          identity {
-              path
-              field
-              format
-          }
-      }
-      automationToken {
-        path
-        field
-        format
-      }
-      internal
-      disable {
-        integrations
-      }
-    }
-  }
-}
-"""
-
 QONTRACT_INTEGRATION = 'openshift-namespaces'
 
+NS_STATUS_PRESENT = "present"
+NS_STATUS_ABSENT = "absent"
+
+NS_ACTION_DELETE = "delete"
+NS_ACTION_CREATE = "create"
+
+
+def get_desired_state(
+        namespaces: Iterable[Mapping[str, Any]]) -> List[Dict[str, str]]:
 
-def get_desired_state(internal, use_jump_host, thread_pool_size):
-    gqlapi = gql.get_api()
-    all_namespaces = gqlapi.query(QUERY)['namespaces']
+    desired_state: List[Dict[str, str]] = []
+    for namespace in namespaces:
+        state = NS_STATUS_PRESENT
+        if namespace.get("delete"):
+            state = NS_STATUS_ABSENT
 
-    namespaces = []
+        desired_state.append({"cluster": namespace["cluster"]["name"],
+                              "namespace": namespace["name"],
+                              "state": state})
+
+    return desired_state
+
+
+def get_shard_namespaces() -> List[Dict[str, str]]:
+    all_namespaces = queries.get_namespaces(minimal=True)
+    namespaces = {}
+    to_be_ignored = set()
     for namespace in all_namespaces:
         shard_key = f'{namespace["cluster"]["name"]}/{namespace["name"]}'
         if is_in_shard(shard_key):
-            namespaces.append(namespace)
+            if shard_key not in namespaces:
+                namespaces[shard_key] = namespace
+            else:
+                to_be_ignored.add(shard_key)
+    for shard_key in to_be_ignored:
+        logging.debug(
```

Comment:
Is this a debugging condition or an error? This code is dropping on the floor these namespaces and, in production, it won't leave any traces of why they were dismissed.

Reply:
OK. I wanted to manage this corner case the same way it's managed in the namespace_labels integration. I understood that this will be controlled with other mechanism and this code will be removed in the near future. I think it makes sense to report error in this case but we should do the same in the other integration as well.

Reply:
The problem is, we don't print debug messages in production, but this is extremely useful when a user has submitted a bad configuraton and needs to understand why.

Reply:
Then, is it ok in this case to print error messages but not fail the integration?

Reply:
I think it makes sense and I've changed it with a slightly consideration. It throws error if there are discrepancies in the namespace definitions. 
- If all the definitions have the delete set, the ns is deleted 
- if all of them don't have the delete set and the namespace does not exist,  it's created ( I think this last one should never happen, but it's covered the same way)